### PR TITLE
DXC Integration in DX12 Backend

### PR DESCRIPTION
**Pull request by 39ali**
**Checklist**

- [x] Run `cargo clippy`.
- [ ] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [ ] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
[https://github.com/gfx-rs/wgpu/issues/2722](https://github.com/gfx-rs/wgpu/issues/2722)

**Description**
DXC will be the default shader compiler for the d3d12 backend , I'm not sure about the addition of `enum ShaderDXIL` , the other solution would be to convert the `ID3DBlob` to `Vec<u8>` which would require a malloc+copy and freeing the blob .  

**Testing**
_Explain how this change is tested._

### CODE REVIEW

File: wgpu-hal/src/dx12/device.rs [hunk 1]
```diff
@@ -221,64 +222,147 @@ impl super::Device {
             .map(|name| ffi::CString::new(name.as_str()).unwrap())
             .map_err(|e| crate::PipelineError::Linkage(stage_bit, format!("{}", e)))?;
 
-        let mut shader_data = native::Blob::null();
-        let mut error = native::Blob::null();
-        let mut compile_flags = d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS;
+        let empty_str = "";
```
Comment (notgull) [inline wgpu-hal/src/dx12/device.rs:225]: Is it necessary to make `empty_str` its own variable here? In fact, you could probably make this entire block:

```rust
let source_name = stage.module.raw_name.as_ref().and_then(|cstr| cstr.to_str()).unwrap_or("");
```

File: wgpu-hal/src/dx12/device.rs [hunk 2]
```diff
@@ -221,64 +222,147 @@ impl super::Device {
...
-                (
-                    Err(crate::PipelineError::Linkage(stage_bit, full_msg)),
-                    log::Level::Warn,
-                )
             }
         };
 
+        // hassle_rs doesn't check for `dxil.dll` if you only call `hassle_rs::compile_hlsl` but calling hassle_rs::Dxil::new will .
+        profiling::scope!("hassle_rs::compile_hlsl");
+        let (result, log_level) = match hassle_rs::Dxil::new(None).and(hassle_rs::compile_hlsl(
```
Comment (notgull) [inline wgpu-hal/src/dx12/device.rs:305]: You may want to make this `and_then` so that `hassle_rs::compile_hlsl` isn't called if `Dxil` fails.

Comment (39ali) [reply wgpu-hal/src/dx12/device.rs:305]: nice catch !

File: wgpu-hal/src/dx12/device.rs [hunk 3]
```diff
@@ -1092,7 +1179,7 @@ impl crate::Device<super::Api> for super::Device {
             },
             bind_group_infos,
             naga_options: hlsl::Options {
-                shader_model: hlsl::ShaderModel::V5_1,
```
Comment (Elabajaba) [inline wgpu-hal/src/dx12/device.rs:1095]: Does bumping this to 6.0 remove support for cards that support shadermodel 5.1?

According to techpowerup's gpu database that would mean removing support for Nvidia's Kepler (700 series, and some laptop 900 series) and earlier, as well as AMD's 1st gen GCN (old 7000 series, which had some cards rebranded and rereleased though the 200 and 300 series, and even some OEM only models in the 400 series and a single OEM only card in the 500 series).

Afaik all of those cards have had their driver support dropped at this point.

Comment (Elabajaba) [reply wgpu-hal/src/dx12/device.rs:1095]: Looked into this a bit, DXC only supports SM6.0 and higher, and FXC only supports up to SM5.1. DXC will upgrade anything it compiles below SM6.0 to SM6.0.

Do you know how other engines handle this, or what happens if you try running SM6.0 code on a SM5.1 gpu?
https://github.com/microsoft/DirectXShaderCompiler/issues/4038 seems like it's probably fine?

Comment (39ali) [reply wgpu-hal/src/dx12/device.rs:1095]: If I'm understanding this correctly according to the (docs)[https://learn.microsoft.com/en-us/windows/win32/direct3d12/hardware-feature-levels], if a device supports `D3D_FEATURE_LEVEL_11_0` it should support shader model `6.0` and `5.1`, if this is incorrect, we can call CheckFeatureSupport(D3D12_FEATURE_SHADER_MODEL..) to check for shader support and choose FXC or DXC accordingly. 
unfortunately i don't have an old GPU to test this .

Comment (39ali) [reply wgpu-hal/src/dx12/device.rs:1095]: It seems that only DXC CLI will upgrade SM5.1 to SM6.0, DXC lib would return an error, this won't be an issue if we take the solution i previously suggested.

Comment (Elabajaba) [reply wgpu-hal/src/dx12/device.rs:1095]: Looks like I was wrong about hardware support. GPU companies seemed to have added support for higher shadermodels in driver updates according to the DXC repo.

```
NVIDIA's r396 drivers (r397.64 and later) provide release mode support for DXIL 1.1 and Shader Model 6.1 on Win10 1709 and later, and experimental mode support for DXIL 1.2 and Shader Model 6.2 on Win10 1803 and later. These drivers also support DXR in experimental mode.

AMD’s driver (Radeon Software Adrenalin Edition 18.4.1 or later) provides release mode support for DXIL 1.1 and Shader Model 6.1.

Intel's 15.60 drivers (15.60.0.4849 and later) support release mode for DXIL 1.0 and Shader Model 6.0 as well as release mode for DXIL 1.1 and Shader Model 6.1 (View Instancing support only).
```

Comment (cwfitzgerald) [reply wgpu-hal/src/dx12/device.rs:1095]: We can assume that all DX12 devices support DXIL - DXBC is already being converted to DXIL by the d3d12 runtime behind the scenes.

File: wgpu-hal/src/dx12/device.rs [hunk 4]
```diff
@@ -221,62 +222,148 @@ impl super::Device {
...
+                "{}_{}\0",
+                naga_stage.to_hlsl_str(),
+                hlsl::ShaderModel::V5_1.to_str()
+            );
+            let mut error = native::Blob::null();
+
+            let hr = unsafe {
+                profiling::scope!("d3dcompiler::D3DCompile");
+                d3dcompiler::D3DCompile(
+                    source.as_ptr() as *const _,
```
Comment (Elabajaba) [inline wgpu-hal/src/dx12/device.rs:265]: Preferable to use .cast() here instead of as casts for all of these pointer casts.

**Review by Elabajaba — CHANGES_REQUESTED** [review]
Clippy errors also need to be fixed, and imo this needs visible docs for users about why they should use dxc, and how they can use it.

File: wgpu-hal/src/dx12/device.rs (hunk 3 above)
Comment (39ali) [reply wgpu-hal/src/dx12/device.rs:1095]: @cwfitzgerald `winapi` only exposes `SM6.0` and `SM5.1` , once wgpu starts using `windows-rs` we can start using the highest supported SM , supporting higher SM is a must so we can support other features like float16,Subgroup etc... ,
Using DXC will still be a big win even if we only support SM6.0 for now .

File: wgpu-hal/src/dx12/device.rs [hunk 5]
```diff
@@ -1482,14 +1579,19 @@ impl crate::Device<super::Api> for super::Device {
...
+                },
                 0,
                 native::CachedPSO::null(),
                 native::PipelineStateFlags::empty(),
             )
         };
 
-        blob_cs.destroy();
+        if let super::ShaderDXIL::FXC(cs_fxc) = blob_cs {
+            cs_fxc.destroy();
```
[the remark this answers is not in the document]
Comment (39ali) [reply wgpu-hal/src/dx12/device.rs:1593]: not sure im getting it , Blob will only be created if FXC was used , when we use DXC no blob will be allocated .

File: wgpu-hal/src/dx12/device.rs [hunk 6]
```diff
@@ -221,62 +222,148 @@ impl super::Device {
...
+                }
+            }
         };
 
-        let hr = unsafe {
-            profiling::scope!("d3dcompiler::D3DCompile");
-            d3dcompiler::D3DCompile(
-                source.as_ptr() as *const _,
-                source.len(),
+        // hassle_rs doesn't check for `dxil.dll` if you only call `hassle_rs::compile_hlsl` but calling hassle_rs::Dxil::new will .
```
[the remark this answers is not in the document]
Comment (39ali) [reply wgpu-hal/src/dx12/device.rs:304]: Good point , as of now `dx12` mod is only used on windows and looking at CI i don't think we validate shaders, let alone DXCIL .

File: wgpu-hal/src/dx12/device.rs (hunk 3 above)
Comment (cwfitzgerald) [reply wgpu-hal/src/dx12/device.rs:1095]: The SM numbers are just enums, we can locally define the higher SMs

File: wgpu-hal/src/dx12/mod.rs [hunk 7]
```diff
@@ -527,6 +527,11 @@ pub struct ShaderModule {
     raw_name: Option<ffi::CString>,
 }
 
+enum ShaderDXIL {
```
Comment (cwfitzgerald) [inline wgpu-hal/src/dx12/mod.rs:530]: This has a misleading name, this is a compiled shader, not DXIL

File: wgpu-hal/Cargo.toml [hunk 8]
```diff
@@ -27,7 +27,7 @@ metal = ["naga/msl-out", "block", "foreign-types"]
 vulkan = ["naga/spv-out", "ash", "gpu-alloc", "gpu-descriptor", "libloading", "smallvec"]
 gles = ["naga/glsl-out", "glow", "egl", "libloading"]
 dx11 = ["naga/hlsl-out", "native", "libloading", "winapi/d3d11", "winapi/d3d11_1", "winapi/d3d11_2", "winapi/d3d11sdklayers", "winapi/dxgi1_6"]
-dx12 = ["naga/hlsl-out", "native", "bit-set", "range-alloc", "winapi/d3d12", "winapi/d3d12shader", "winapi/d3d12sdklayers", "winapi/dxgi1_6"]
+dx12 = ["naga/hlsl-out", "native", "bit-set", "range-alloc", "winapi/d3d12", "winapi/d3d12shader", "winapi/d3d12sdklayers", "winapi/dxgi1_6", "hassle-rs"]
```
Comment (cwfitzgerald) [inline wgpu-hal/Cargo.toml:30]: We probably shouldn't be depending on this unconditionally. It depends on the needs of firefox.

@jimblandy tl;dr: hassle-rs are bindings to DXC, they're not terribly thick bindings, but they exist. If FF is going to ship on FXC, do you need a way to disable support DXC entirely?

File: wgpu-hal/src/dx12/device.rs [hunk 9]
```diff
@@ -221,62 +222,151 @@ impl super::Device {
...
-                            error.GetBufferPointer() as *const u8,
-                            error.GetBufferSize(),
-                        )
-                    };
-                    let _ = write!(full_msg, ": {}", String::from_utf8_lossy(message));
-                    unsafe {
-                        error.destroy();
-                    }
+                hassle_rs::HassleError::LoadLibraryError { filename, inner: _ } => {
+                    log::warn!("{:?} was not found ,falling back to FXC ,in order to use the new hlsl compiler `DXC`,`dxcompiler.dll` and `dxil.dll` should be included in the current execution environment" ,filename );
```
Comment (cwfitzgerald) [inline wgpu-hal/src/dx12/device.rs:344]: We should only send this warning once per program execution.

File: wgpu-hal/src/dx12/device.rs [hunk 10]
```diff
@@ -221,62 +222,151 @@ impl super::Device {
...
-        let mut error = native::Blob::null();
-        let mut compile_flags = d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS;
+        let source_name = stage
+            .module
+            .raw_name
+            .as_ref()
+            .and_then(|cstr| cstr.to_str().ok())
+            .unwrap_or_default();
+
+        let mut compile_flags = vec!["-Ges"]/* d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS */;
```
Comment (cwfitzgerald) [inline wgpu-hal/src/dx12/device.rs:232]: Lets use an ArrayVec<_, 3> here

File: wgpu-hal/src/dx12/device.rs [hunk 11]
```diff
@@ -221,62 +222,151 @@ impl super::Device {
...
+            compile_flags.push("-Od"); /* d3dcompiler::D3DCOMPILE_DEBUG */
         }
 
-        let source_name = match stage.module.raw_name {
-            Some(ref cstr) => cstr.as_c_str().as_ptr(),
-            None => ptr::null(),
+        // This closure gets called if `hassle_rs::compile_hlsl() fails due to a `hassle_rs::HassleError::LoadLibraryError`,
+        //
+        //  This usually means that `dxcompiler.dll` and/or `dxil.dll` could not be found in the current execution environment.
+        let load_shader_fxc = || {
```
Comment (cwfitzgerald) [inline wgpu-hal/src/dx12/device.rs:245]: Having this much code in a closure is a bit of a code smell. It should likely be its own function.

File: wgpu-hal/src/dx12/device.rs [hunk 12]
```diff
@@ -221,62 +222,151 @@ impl super::Device {
...
 
-        let hr = unsafe {
-            profiling::scope!("d3dcompiler::D3DCompile");
-            d3dcompiler::D3DCompile(
-                source.as_ptr() as *const _,
-                source.len(),
+        // hassle_rs doesn't check for `dxil.dll` if you only call `hassle_rs::compile_hlsl` but calling hassle_rs::Dxil::new will .
+        profiling::scope!("hassle_rs::compile_hlsl");
+        let (result, log_level) = match hassle_rs::Dxil::new(None).and_then(|_| {
+            hassle_rs::compile_hlsl(
```
Comment (cwfitzgerald) [inline wgpu-hal/src/dx12/device.rs:310]: We shouldn't be using `compile_hlsl` directly as it needs to load the library every single time you call the function. Ideally you'll load a DXCCompiler and DXCLibrary once on dx12 instance creation, then re-use it for every time we need to compile stuff.

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]

File: wgpu-hal/src/dx12/device.rs (hunk 9 above)
Comment (MarijnS95) [reply wgpu-hal/src/dx12/device.rs:344]: And what's up with the commas?

**Comment by cwfitzgerald** [comment]
Superceded by #3356

**Pull request closed without merging**